### ISSUE TITLE
Merging namespace from different modules

### ISSUE THREAD

**Issue by darkbasic** [Microsoft/TypeScript#18767]
Hi,
I use a module called [`node-simple-schema`](https://github.com/aldeed/node-simple-schema) which adds a property called `schema` on top of Meteor's [`Mongo.Collection`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/meteor/mongo.d.ts).

`node-simple-schema` allows me to do:
```
import { Mongo } from 'meteor/mongo';
export const Components = new Mongo.Collection<Component>('components');
Components.schema = new SimpleSchema<Component>({});
```

This is how I created working type definitions (`my-typings.d.ts`):
```
declare module 'meteor/mongo' {
  import SimpleSchema from 'simpl-schema';

  namespace Mongo {
    interface Collection<T> {
      schema?: SimpleSchema<T>;
    }
  }
}
```

I would like to switch from `Mongo` to [`MongoObservable`](https://github.com/Urigo/meteor-rxjs/blob/master/dist/ObservableCollection.d.ts):
```
import { MongoObservable } from "meteor-rxjs";
export const ComponentsReactive = new MongoObservable.Collection<Component>('components');
ComponentsReactive.collection.schema = new SimpleSchema<Component>({});
```

Objects of type `MongoObservable.Collection` have a `collection` property of type `Mongo.Collection<T>`, which allows me to expose the `schema` property. This is how I expected type definitions (`my-typings.d.ts`) to be like:
```
declare module "meteor-rxjs" {
  import SimpleSchema from 'simpl-schema';
  namespace Mongo {
    interface Collection<T> {
      schema?: SimpleSchema<T>;
    }
  }

  namespace MongoObservable {
    interface Collection<T> {
      readonly collection: Mongo.Collection<T>;
    }
  }
}
```

Unfortunately with such type definitions I get the error `TS2708: Cannot use namespace 'MongoObservable' as a value.`
It seems like if they overridden the namespace instead of merging it.

Instead the following type definitions (`my-typings.d.ts`) work flawlessly:
```
import SimpleSchema from 'simpl-schema';

declare global {
  namespace Mongo {
    interface Collection<T> {
      schema?: SimpleSchema<T>;
    }
  }

  namespace MongoObservable {
    interface Collection<T> {
      readonly collection: Mongo.Collection<T>;
    }
  }
}
```

Even if it works, I don't like the idea of polluting the global namespace, especially if I plan to release my type definitions.
What I would really like to know is why I have been able to merge namespaces from different modules in the first case while it didn't work in the second case.
Also I would like to know what's the correct way to achieve my goal: creating good type definitions for `node-simple-schema` and `MongoObservable`.

Thanks

**Comment by mhegazy** [comment]
Modules have thier own scope. so a namespace within a module does not merge with a namespace with the same name in the global scope.

seems like `Mango` is declared as a global anyways, so do not see how you can use it and augment it without having your declarations in the same scope.

**Comment by darkbasic** [comment]
Yes, `Mongo` is also declared as a global, but the global declaration is only a subset of `meteor/mongo`, so most of the times I still need to `import { Mongo } from 'meteor/mongo'` anyway.

Anyway let's leave `Mongo` alone for a moment and let's say I want to augment `MongoObservable`:
```
declare module "meteor-rxjs" {
  namespace MongoObservable {
    interface Collection<T> {
      something: any;
    }
  }
}
```
```
import { MongoObservable } from "meteor-rxjs";
export const ComponentsReactive = new MongoObservable.Collection<Component>('components');
```

Why does it lead to the error `TS2708: Cannot use namespace 'MongoObservable' as a value`?

**Comment by mhegazy** [comment]
Do you have a self-contained repro i can look at?

**Comment by mhegazy** [comment]
When augmenting a declaration originating from a module the augmentation has to be on the source of the declaarations, and not the reexports.

`MangoObservable.Collection` comes from `"meteor-rxjs/dist/ObservableCollection"` and not `"meteor-rxjs"`. 

The other issue is that a module augmentation is always nested in anohter module. otherwise it is a declration that shadows the other module.

so, your augmentation should be something like:

````ts

/// charts.ts

import { MongoObservable } from "meteor-rxjs";
export const ChatsRxjs = new MongoObservable.Collection('chats');

ChatsRxjs.something;


declare module "meteor-rxjs/dist/ObservableCollection" {
        namespace MongoObservable {
            interface Collection<T> {
                something: any;
            }
        }
    }
```

**Comment by darkbasic** [comment]
> When augmenting a declaration originating from a module the augmentation has to be on the source of the declaarations, and not the reexports.

Thanks, do you have a source for it? Becuase I searched a lot before asking here and I didn't find anything about it.
I don't like the idea of having to refer to the exact file name, what if they refactor their code and the file name changes?

With your suggestion my original type definition becomes:
```
import SimpleSchema from 'simpl-schema';

declare module "meteor-rxjs/dist/ObservableCollection" {
  namespace Mongo {
    interface Collection<T> {
      attachSchema(schema: SimpleSchema<T>, opts?: {
        selector?: any;
        transform?: boolean;
        replace?: boolean;
      }): void;
    }
  }

  namespace MongoObservable {
    interface Collection<T> {
      readonly collection: Mongo.Collection<T>;
    }
  }
}
```

Unfortunately my problems are far from being over, let's see what happens when I need to create type definitions for something bigger like `MongoObservable` + `meteor-collection-hooks`:

```
import { Mongo as myMongo } from 'meteor/mongo';

declare module "meteor-rxjs/dist/ObservableCollection" {
  namespace Mongo {
    interface Collection<T> {
      before: {
        find(hook: { (userId: string, selector: myMongo.Selector, options: { multi?: boolean; upsert?: boolean; }): void }): void;
        findOne(hook: { (userId: string, selector: myMongo.Selector, options: { multi?: boolean; upsert?: boolean; }): void }): void;
        insert(hook: { (userId: string, doc: T): void }): void;
        remove(hook: { (userId: string, doc: T): void }): void;
        update(hook: { (userId: string, doc: T, fieldNames: string[], modifier: myMongo.Modifier, options: { multi?: boolean; upsert?: boolean; }): void }): void;
        upsert(hook: { (userId: string, doc: T, selector: myMongo.Selector, modifier: myMongo.Modifier, options: { multi?: boolean; upsert?: boolean; }): void }): void;
      };
      after: {
        find(hook: { (userId: string, selector: myMongo.Selector, options: { multi?: boolean; upsert?: boolean; }, cursor: myMongo.Cursor<T>): void }): void;
        findOne(hook: { (userId: string, selector: myMongo.Selector, options: { multi?: boolean; upsert?: boolean; }, doc: T): void }): void;
        insert(hook: { (userId: string, doc: T): void }): void;
        remove(hook: { (userId: string, doc: T): void }): void;
        update(hook: { (userId: string, doc: T, fieldNames: string[], modifier: myMongo.Modifier, options: { multi?: boolean; upsert?: boolean; }): void }): void;
        upsert(hook: { (userId: string, doc: T, selector: myMongo.Selector, modifier: myMongo.Modifier, options: { multi?: boolean; upsert?: boolean; }): void }): void;
      };
      direct: {
        find(selector?: myMongo.Selector | myMongo.ObjectID | string, options?: {
          sort?: myMongo.SortSpecifier;
          skip?: number;
          limit?: number;
          fields?: myMongo.FieldSpecifier;
          reactive?: boolean;
          transform?: Function;
        }): myMongo.Cursor<T>;
        findOne(selector?: myMongo.Selector | myMongo.ObjectID | string, options?: {
          sort?: myMongo.SortSpecifier;
          skip?: number;
          fields?: myMongo.FieldSpecifier;
          reactive?: boolean;
          transform?: Function;
        }): T;
        insert(doc: T, callback?: Function): string;
        remove(selector: myMongo.Selector | myMongo.ObjectID | string, callback?: Function): number;
        update(selector: myMongo.Selector | myMongo.ObjectID | string, modifier: myMongo.Modifier, options?: {
          multi?: boolean;
          upsert?: boolean;
        }, callback?: Function): number;
        upsert(selector: myMongo.Selector | myMongo.ObjectID | string, modifier: myMongo.Modifier, options?: {
          multi?: boolean;
        }, callback?: Function): { numberAffected?: number; insertedId?: string; };
      };
      hookOptions: CollectionOptions;
    }
  }

  namespace MongoObservable {
    interface Collection<T> {
      readonly collection: Mongo.Collection<T>;
    }
  }
}
```

Most of the `Mongo` types I need are not available on the global scope, so I need to import `Mongo` explicitely, but I also have to import it **as** something else (`myMongo` in my example) because otherwise it will collide with my `namespace Mongo {}`. Then I need to paste a great amount of code from the `meteor-collection-hooks` type definitions and manually change every occurrency of `Mongo` with `myMongo`. Also, every time something changes in `meteor-collection-hooks` I will have to rewrite the whole type definition.

I would like to be able to automatically import the namespace from available type definitions, something like:
```
declare module "meteor-rxjs/dist/ObservableCollection" {
  namespace Mongo from "meteor/mongo" {}

  namespace MongoObservable {
    interface Collection<T> {
      readonly collection: Mongo.Collection<T>;
    }
  }
}
```

Is it possible to achieve something like that?

Last but not the least, since I want to make my type definitions available to everyone on the `@types` namespace, which package am I supposed to upload them to? `@types/mongo-observable` sounds like a good candidate, but `meteor-collection-hooks` is a completely optional package so I would say no. `@types/meteor-collection-hooks`? Again, `MongoObservable` is completely optional, so...

Thanks for your great help

**Comment by darkbasic** [comment]
I found even another issue. When using `declare module "meteor-rxjs/dist/ObservableCollection"` instead of `declare global` the following:

```
import SimpleSchema from 'simpl-schema';

declare module "meteor-rxjs/dist/ObservableCollection" {
  namespace Mongo {
    interface Collection<T> {
      attachSchema(schema: SimpleSchema<T>, opts?: {
        selector?: any;
        transform?: boolean;
        replace?: boolean;
      }): void;
    }
  }

  namespace MongoObservable {
    interface Collection<T> {
      readonly collection: Mongo.Collection<T>;
    }
  }
}
```
overrides my `Mongo` namespace for `MongoObservable` instead of merging it!

In fact Typescript complains that property `find` does not exist on type `Collection`:
```
export const ChatsReactive = new MongoObservable.Collection('chats');
export const Chats = ChatsReactive.collection;
Chats.find();
```

It didn't happen when I was using `declare global`, so unfortunately this isn't a solution.

**Comment by mhegazy** [comment]
Automatically closing this issue for housekeeping purposes. The issue labels indicate that it is unactionable at the moment or has already been addressed.

**Comment by darkbasic** [comment]
Hi,
Can you please re-open? The issue has not being already addressed. What do you mean with unactionable? That it isn't possible to create type definitions for this library?
